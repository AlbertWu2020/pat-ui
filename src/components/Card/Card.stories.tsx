import React from 'react';
import { action } from '@storybook/addon-actions';
import Card from './Card';

export default {
  title: 'Card',
  component: Card,
};

export const DefaultCard = () => (
  <Card
    onClick={action('Default Button clicked')}
    src={'https://via.placeholder.com/150'}
  >
    Default Card
  </Card>
);

export const DiffSizeCard = () => (
  <div>
    <Card
      onClick={action('Default Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardSize='sm'
    >
      Small Card
    </Card>

    <Card
      onClick={action('Default Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardSize='lg'
    >
      Large Card
    </Card>
  </div>
);
export const DiffTypeCard = () => (
  <div>
    <Card
      onClick={action('Default Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardType='horizontal'
    >
      Horizontal Card
    </Card>
    <Card
      onClick={action('Default Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardType='circledImage'
    >
      Circled Image Card
    </Card>
    <Card
      onClick={action('Default Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardType='noImage'
    >
      No Image Card
    </Card>
    <Card
      onClick={action('Default Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardType='largeImage'
    >
      Large Image Card
    </Card>
  </div>
);
export const DiffThemeCard = () => (
  <div>
    <Card
      onClick={action('Dark Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardTheme='dark'
    >
      Dark Theme Card
    </Card>

    <Card
      onClick={action('Purple Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardTheme='purple'
    >
      Purple Theme Card
    </Card>
    <Card
      onClick={action('Yellow Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardTheme='yellow'
    >
      Yellow Theme Card
    </Card>
    <Card
      onClick={action('Blue Button clicked')}
      src={'https://via.placeholder.com/150'}
      cardTheme='blue'
    >
      Blue Theme Card
    </Card>
  </div>
);
