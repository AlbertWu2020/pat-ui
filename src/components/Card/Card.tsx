import React, { FC } from 'react';
import Button from '../Button/Button';
import { classNames } from '../../utils/classNames';

export type CardSize = 'lg' | 'sm';
export type CardMode = 'default' | 'horizontal';
export type CardTheme = 'dark' | 'purple' | 'blue' | 'yellow';
export type CardType = 'circledImage' | 'noImage' | 'largeImage';

export interface ICardProps {
  /** set customized card */
  className?: string;
  /** set card content */
  cardParagraph?: string;
  /** set card title */
  cardTitle?: string;
  /** set card image source */
  cardImgSrc?: string;
  /** set card theme */
  cardTheme?: CardTheme;
  /** set card mode */
  cardMode?: CardMode;
  /** set card size */
  cardSize?: CardSize;
  /** set card type */
  cardType?: CardType;
  /** set action on bottun clicked */
  btnOnClick?: () => void;
  /** set the redirect page on button clicked*/
  buttonHref?: string;
  /** set button title */

  buttonTitle?: string;
}
// export interface ImgHTMLAttributes<T> {
//   src?: string;
// }

export type patCardProps = ICardProps;
/**
 * A card with a button to show more
 *
 */
export const Card: FC<patCardProps> = (props) => {
  const {
    cardSize,
    cardType,
    cardMode,
    cardTheme,
    children,
    className,
    buttonHref,
    buttonTitle,
    ...rest
  } = props;
  let styleClasses = classNames('card', {
    [`card-${cardMode}`]: true,
    [`card-${cardType}`]: true,
    [`card-${cardTheme}`]: true,
    [`card-${cardSize}`]: !!cardSize,
  });
  if (className) {
    styleClasses += ' ' + className;
  }

  let Card = (
    <div className={styleClasses} data-testid='card-element'>
      <div className={styleClasses + ' image'} data-testid='card-image-element'>
        <img src={props.cardImgSrc} data-testid='image-element' />
      </div>
      <div className={styleClasses + ' body'} data-testid='card-body-element'>
        <h5>{props.cardTitle}</h5>
        <p>{props.cardParagraph}</p>
        <Button
          onClick={props.btnOnClick}
          data-testid='button-element'
          className='btn btn-primary'
          href={props.buttonHref}
        >
          {props.buttonTitle}
        </Button>
      </div>
    </div>
  );
  return Card;
};
Card.defaultProps = {
  cardMode: 'default',
};
export default Card;
