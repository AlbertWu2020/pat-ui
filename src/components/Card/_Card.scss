.card-default {
  @include card-border;
  width: 15rem;
  display: flex;
  flex-direction: column;
  .image {
    margin: 0px;
  }
  .image img {
    height: 10rem;
  }
  .body {
    @include card-body-style;
  }
  .btn {
    width: 10rem;
  }
}

.card-horizontal {
  @include card-border;
  width: 40rem;
  height: 10rem;
  display: flex;
  flex-direction: row;
  .image {
    margin: 0px;
    width: 10rem;
  }
  .body {
    @include card-body-style;
    display: flex;
    flex-direction: column;
  }
  .btn {
    width: 10rem;
  }
}
.card-circledImage {
  @include card-border;
  margin-right: '5px';
  width: 15rem;
  display: flex;
  flex-direction: column;
  .image {
    margin: 0px;
    height: 10rem;
    display: flex;
    border-style: none;
    padding: 0.5rem;
  }
  .image img {
    width: 10rem;
    border-radius: 50%;
    align-self: center;
  }
  .body {
    @include card-body-style;
  }
  .btn {
    width: 10rem;
  }
}

.card-noImage {
  @include card-border;
  width: 20rem;
  display: flex;
  flex-direction: column;
  .image {
    margin: 0px;
    display: none;
  }
  .body {
    @include card-body-style;
  }
  .btn {
    width: 10rem;
  }
}

.card-dark {
  @include card-border;
  width: 15rem;
  display: flex;
  flex-direction: column;
  .image {
    margin: 0px;
  }
  .image img {
    height: 10rem;
  }
  .body {
    @include card-body-style;
    margin: 0px;
    background-color: $gray-800;
    color: $white;
  }
  .btn {
    @include button-style($white, $secondary, $secondary);
    width: 10rem;
  }
}
.card-purple {
  @include card-theme-style($purple);
}
.card-blue {
  @include card-theme-style($blue);
}

.card-yellow {
  @include card-theme-style($dark-yellow);
  .btn {
    @include button-style($white, $dark-yellow, $yellow);
    width: 10rem;
  }
}

.card-sm {
  @include card-border;
  width: 12rem;
  display: flex;
  flex-direction: column;
  .image {
    margin: 0px;
  }
  .image img {
    height: 8rem;
  }
  .body {
    @include card-body-style;
  }
  .btn {
    width: 8.5rem;
  }
}

.card-lg {
  @include card-border;
  width: 18rem;
  height: 25rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  .image {
    margin: 0px;
  }
  .image img {
    height: 12rem;
  }
  .body {
    @include card-body-style;
    font-size: $font-size-base;
  }
  .body h5 {
    font-size: 1.5rem;
  }
  .btn {
    width: 12rem;
  }
}

.card-largeImage {
  @include card-border;
  width: 18rem;
  display: flex;
  flex-direction: column;
  .image {
    margin: 0px;
  }
  .image img {
    height: 12rem;
  }
  .body {
    text-align: center;
    @include card-body-style;
  }
  .body p {
    display: none;
  }
  .btn {
    display: none;
  }
}
