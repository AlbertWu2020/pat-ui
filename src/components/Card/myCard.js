import React from 'react';
import Button from '../Button/Button';
function MyCard(props) {
  return (
    <div className={props.className} data-testid='card-element'>
      <div
        className={props.className + ' image'}
        data-testid='card-image-element'
      >
        <img src={props.src} data-testid='image-element' />
      </div>
      <div
        className={props.className + ' body'}
        data-testid='card-body-element'
      >
        <h5>{props.children}</h5>
        <p>
          Some quick example text to build on the card title and make up the
          bulk of the card's content.
        </p>
        <Button
          onClick={props.onClick}
          data-testid='button-element'
          className='btn btn-primary'
          href='#'
        >
          Show More Info
        </Button>
      </div>
    </div>
  );
}

export default MyCard;
