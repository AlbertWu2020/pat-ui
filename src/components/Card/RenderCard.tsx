import React from "react";
import { Card } from "../Card/Card";
import { CardSize, CardType } from "./Card";
export function RenderCard() {
  return (
    <>
      <h2>Card Component</h2>
      <h2>Default</h2>
      <Card cardType={CardType.Default}></Card>
      <h2>Horizontal</h2>
      <Card cardType={CardType.Horizontal}></Card>
      <h2>CircledImage</h2>
      <Card cardType={CardType.CircledImage}></Card>
      <h2>NoImage</h2>
      <Card cardType={CardType.NoImage}></Card>
      <h2>Dark</h2>
      <Card cardType={CardType.Dark}></Card>
    </>
  );
}
