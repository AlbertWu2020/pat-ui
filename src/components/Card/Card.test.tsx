import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import Card, { myCardProps } from './Card';
describe('Card', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Card> Snapshot Card </Card>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render default card', () => {
    const cardProps = {
      onClick: jest.fn(),
    };
    const wrapper = render(<Card {...cardProps}></Card>);
    const cardElement = screen.getByTestId('card-element');
    expect(cardElement).toHaveClass('card card-default');
  });
  it('should render dark card with correct title', () => {
    const cardProps: myCardProps = {
      cardTheme: 'dark',
    };
    const wrapper = render(<Card {...cardProps}>Dark Card</Card>);
    //test h5 title
    const titleElement = wrapper.queryByText('Dark Card') as HTMLElement;
    expect(titleElement).toBeInTheDocument();
    expect(titleElement.tagName).toBe('H5');
    //test class name
    const cardElement = screen.getByTestId('card-element');
    expect(cardElement).toHaveClass('card card-dark');
  });
  it('should render card on different props', () => {
    const cardProps: myCardProps = {
      cardSize: 'lg',
      cardType: 'noImage',
      className: 'test',
    };
    const wrapper = render(<Card {...cardProps}>noImage Card</Card>);
    //test h5 title
    const titleElement = wrapper.queryByText('noImage Card') as HTMLElement;
    expect(titleElement).toBeInTheDocument();
    //test class name
    const cardElement = screen.getByTestId('card-body-element');
    expect(cardElement).toHaveClass('card card-noImage card-lg body test');
    const cardBodyElement = screen.getByTestId('card-image-element');
    expect(cardBodyElement).toHaveClass('card card-noImage card-lg image test');
  });

  it('should render button can be clicked', () => {
    const cardProps = {
      onClick: jest.fn(),
    };
    //test button
    const wrapper = render(<Card {...cardProps}></Card>);
    const buttonElement = screen.getByTestId('button-element');
    expect(buttonElement.tagName).toBe('BUTTON');
    expect(cardProps.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(buttonElement);
    expect(cardProps.onClick).toHaveBeenCalledTimes(1);
  });
  it('should render image with correct source', () => {
    const cardProps = {
      src: 'png',
    };
    const wrapper = render(<Card {...cardProps}></Card>);
    const imageElement = screen.getByTestId('image-element');
    expect(imageElement.tagName).toBe('IMG');
    expect(imageElement.getAttribute('src')).toBe('png');
  });
});
