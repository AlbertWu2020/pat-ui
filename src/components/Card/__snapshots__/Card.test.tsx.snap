// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="card card-default"
    data-testid="card-element"
  >
    <div
      class="card card-default image"
      data-testid="card-image-element"
    >
      <img
        data-testid="image-element"
      />
    </div>
    <div
      class="card card-default body"
      data-testid="card-body-element"
    >
      <h5>
         Snapshot Card 
      </h5>
      <p>
        Some quick example text to build on the card title and make up the bulk of the card's content.
      </p>
      <button
        class="btn btn-default btn btn-primary"
        data-testid="button-element"
        href="#"
      >
        Show More Info
      </button>
    </div>
  </div>
</DocumentFragment>
`;
